ykbh/src/lib.rs

```diff
@@ -0,0 +1,273 @@
+            let size = SIR.ty(&d.ty).size();
+            let l = Layout::from_size_align(size.try_into().unwrap(), align.try_into().unwrap()).unwrap();
+            let (nl, s) = layout.extend(l).expect("Alignment error");
+            offsets.push(s);
+            layout = nl;
+        }
+
+        // Allocate memory for the locals
+        let locals = unsafe { alloc(layout.pad_to_align()) };
+
+        SIRInterpreter {
+            locals,
+            offsets,
+            layout,
+        }
+    }
+
+    /// Inserts a pointer to the trace inputs into `locals`.
+    pub fn set_trace_inputs(&mut self, tio: *mut usize) {
+        // FIXME Later this also sets other already initialised variables as well as the program
+        // counter of the interpreter.
+        let ptr = self.local_ptr(&Local(1)); // The trace inputs live in $1
+        unsafe {
+            // Write the pointer value of `tio` into locals by casting it to an integer first.
+            std::ptr::write::<usize>(ptr as *mut usize, tio as usize);
+        }
+    }
+
+    pub fn interpret(&mut self, body: &ykpack::Body) {
+        // Ignore yktrace::trace_debug.
+        if body.flags & ykpack::bodyflags::TRACE_DEBUG != 0 {
+            return;
+        }
+
+        for stmt in body.blocks[0].stmts.iter() {
+            match stmt {
+                Statement::MkRef(dest, src) => self.mkref(dest, src),
+                Statement::DynOffs { .. } => todo!(),
+                Statement::Store(dest, src) => self.store(dest, src),
+                Statement::BinaryOp { .. } => todo!(),
+                Statement::Nop => {}
+                Statement::Unimplemented(_) | Statement::Debug(_) => todo!(),
+                Statement::Cast(..) => todo!(),
+                Statement::Call(..) | Statement::StorageDead(_) => unreachable!(),
+            }
+        }
+
+        match &body.blocks[0].term {
+            Terminator::Call {
+                operand: _op,
+                args: _args,
+                destination: _dest,
+            } => todo!(),
+            Terminator::Return => {}
+            t => todo!("{}", t),
+        }
+    }
+
+    /// Get the pointer to a Local.
+    fn local_ptr(&self, local: &Local) -> *mut u8 {
+        let offset = self.offsets[usize::try_from(local.0).unwrap()];
+        unsafe { self.locals.add(offset) }
+    }
+
+    /// Get the pointer for an IPlace, while applying all offsets.
+    fn iplace_to_ptr(&self, place: &IPlace) -> *mut u8 {
+        match place {
+            IPlace::Val {
+                local,
+                off,
+                ty: _ty,
+            } => {
+                // Get a pointer to the Val.
+                let dest_ptr = self.local_ptr(&local);
+                unsafe { dest_ptr.add(usize::try_from(*off).unwrap()) }
+            }
+            IPlace::Indirect { ptr, off, ty: _ty } => {
+                // Get a pointer to the Indirect.
+                let dest_ptr = self.local_ptr(&ptr.local) as *mut usize;
```

Comment:
Why are we casting this `*mut usize`?

Reply:
We are doing this because `local_ptr` returns a pointer which we are going to dereference below. So casting this to `*mut usize` allows me to do:
```rust
let val = read::<usize>(ptr);
```
instead of something like
```rust
let val = 0;
copy(ptr, &val as *mut u8, size_of<usize>());
```